### ISSUE TITLE
Question about nn.nLLLoss() and  nn.CrossEntropyLoss() 

### ISSUE THREAD

**Issue by fangfchen** [CharlesNord/centerloss#1]
Hi, I find the code uses F.log_softmax() and nn.nLLLoss() to count the softmax entropy loss.
And I use nn.CrossEntropyLoss() instead.
But the result is very bad ——test acc is only  93%. 
Could you please tell me what's wrong? Thank you!

**Comment by CharlesNord** [comment]
> Hi, I find the code uses F.log_softmax() and nn.nLLLoss() to count the softmax entropy loss.
> And I use nn.CrossEntropyLoss() instead.
> But the result is very bad ——test acc is only 93%.
> Could you please tell me what's wrong? Thank you!

我试了下，没有问题，model 里面去掉 F.log_softmax() 然后把 nn.NLLLoss() 改成 nn.CrossEntropyLoss() 后，效果也不错，代码中的卷积网络用了 14 个 epochs 精度达到 99%，线性网络用了 24 个 epochs 精度达到 99%

**Comment by fangfchen** [comment]
我看到代码里面test的构造是这样的：
testset = torchvision.datasets.MNIST(root='../mnist', train=True, transform=transforms.Compose(
    [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
))
所以想问这里用来test是否也是训练集？99%的效果是训练集还是测试集上的呢？谢谢

**Comment by CharlesNord** [comment]
> 我看到代码里面test的构造是这样的：
> testset = torchvision.datasets.MNIST(root='../mnist', train=True, transform=transforms.Compose(
> [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
> ))
> 所以想问这里用来test是否也是训练集？99%的效果是训练集还是测试集上的呢？谢谢

抱歉抱歉，我写代码的时候直接复制粘贴了上段代码，忘了改。刚刚重新做了一下试验，线性网络精度上限是97%，卷积网络还是 99%，F.log_softmax()+nn.NLLLoss() 和直接用 nn.CrossEntropyLoss() 没有明显区别，我的几次实验结果，后者都比前者略好零点几个百分点，不知道是不是随机的。这段代码主要是可视化一下聚类过程，所以bottleneck的feature维度设置成了2，如果设得大一点，效果应该会更好。

**Comment by fangfchen** [comment]
嗯嗯，我刚刚又尝试重新同时跑了这两种写法，发现最后的结果是一样的(acc98.8%)，看来可能是比较受随机因素影响-. -
另外我看这个是每次都会算batch内的center，那batchsize的大小会有影响么？（比如我现在想跑的数据，分类数比batchsize还大。。）

**Comment by CharlesNord** [comment]
> 嗯嗯，我刚刚又尝试重新同时跑了这两种写法，发现最后的结果是一样的(acc98.8%)，看来可能是比较受随机因素影响-. -
> 另外我看这个是每次都会算batch内的center，那batchsize的大小会有影响么？（比如我现在想跑的数据，分类数比batchsize还大。。）

你可以试一试，这是玄学= =。这个算法没有每次计算batch的center，而是根据batch的feature与当前center的距离计算损失，然后根据损失更新center，所以center其实是一步一步更新的。但batchsize小了，center更新会变慢，所以可能 cross-entropyloss 会占主导
